### PR TITLE
Edit posts

### PR DESCRIPTION
**Pull request by dominictracey**
Seemed like a workable item on the roadmap to use to look at presenatational vs. container components and to do a little investigation on state management. I changed some components from presentational to component and vice versa so that I could reuse the editing component both for post creation and post editing. 

The old component was called ReplyToThread, which largely became the EditPost component which is used both in Post, which got a tiny bit of state to manage the toggling between viewing and editing, and in the stripped down ReplyToThread.

I also updated the database security JSON. I haven't been super assiduous is asserting it's veracity but it seems a reasonably good approximation.

In terms of themes and styles, I still have a limited grasp of the big picture there so probably need to so some work around how FreeStyle works and the architecture. I fooled around with making admin buttons reddish. So admins can edit all posts but only other peoples' posts will give them the red edit icon button.

Speaking of icons, I had to take the plus icon out of the ReplyToThread component because I couldn't immediately see a good way of swapping between the plus and pencil(?) icons on the ok button. Only editing a post gets a cancel button.

There may have been a bug in the code with the ReplyTo field for posts not getting populated properly that should be fixed now for both reply and quote. One thing I'd like to see on the roadmap is a nested display for threads, a la reddit - such that responses are shown indented below their parent post.

### CODE REVIEW

File: src/Thread/ReplyToThread.js [hunk 1]
```diff
@@ -1,116 +1,42 @@
...
-            onClick={this.submit}
-          >
-            <PlusIcon className={styles.plusIcon} /> Reply to thread
-          </Button>
-          <PreviewButton
-            enabled={this.state.previewEnabled}
-            togglePreview={this.togglePreview}
-          />
-        </Form>
+        <EditPost
```
Comment (hoppula) [inline src/Thread/ReplyToThread.js:22]: User's profile image is now missing.

**Comment by hoppula** [comment]
Thanks again for yet another PR!

This is looking quite good overall, one thing I noticed is that user avatar images are now missing in thread view, so they could be added back.

I think the styles are fine like this, I'm gonna check how the dark theme looks with these changes later and I can also fix the save button's plus/edit icon.

Good thing that you fixed the `replyTo` mapping, I added it in the beginning just for your proposed use case, nested replies view. 

I've got two things in my mind that I'd like to see added to post data after editing: 1) `edited` boolean flag, which can be used to show that the post has been edited and 2) `editedAt` timestamp, so we have the option to show the edit time.

**Comment by hoppula** [comment]
And we could possibly add `lastEditedBy` field to post as well, it would have editing user's key as value.

**Comment by dominictracey** [comment]
Sounds good. Election night here so not much getting done!

**Comment by dominictracey** [comment]
Ok - this should be set now. I had to revamp the security rules quite a bit so we should probably do some sort of audit on them.

**Comment by hoppula** [comment]
Thanks, this is looking quite good now, though I'm not sure about all the security rule changes, I'll have to review them with time, probably tomorrow.

**Comment by dominictracey** [comment]
It was why I didn't append the edit data into the post objects in the first place, I tried to do it and was getting lots of permission errors. After reading up on the security rules design principles and playing around with the simulator it seemed like the best (only?) way to do it was to validate two types of writes: the initial creation and an edit. The edit seems to need to include all of the post's data, not just the updated/new fields. If I didn't include the (e.g.) threadId, it was deleted by the edit write.

The overriding learning for me was that write rules don't seem to combine. So I think we do need to have an overarching write for authenticated users and then implement various requirements using validate() calls within the object hierarchy. I tried to put some comments into the security-rules.json describing what requirements were being implemented where.

**Comment by hoppula** [comment]
I think that the current `security-rules.json` has gotten so complex that it's best to switch to Bolt:
https://github.com/firebase/bolt

I'll try to migrate the current `security-rules.json` to simplified .bolt format with comments and you can then rebase this PR to use those security rules and add new rules in Bolt format.

Bolt allows us to define security rules as functions and TypeScript types, I think it'll be a big improvement over current manual rule management. This example seems especially relevant:
https://github.com/firebase/bolt/blob/master/samples/user-security.bolt

**Comment by dominictracey** [comment]
This makes sense to me. It did make me nervous how difficult to read and debug firebase rules was. Anything that can allow us to get from business requirements to rules is a Good Thing. 

As an FYI, I'm pushing ahead with implementing different ways to start threads (e.g. posting a link a la Reddit or Slashdot). I have a working version deployed you can look at [here](https://rugby-forum-poc.firebaseapp.com/board/0), but am going to take some time to rethink the implementation to allow extension for other types of "content", besides self (text) and link, to initiate threads. For our use case, we'll want to be able to have the thread header be rendered differently for different content types. In the original code, it is only the thread title. With the new link threads, it's title, image and description. I'd like to provide a proscribed extension methodology for creating threads from other content types.

**Comment by hoppula** [comment]
I just pushed security-rules generated with Bolt, they're now defined in [security-rules.bolt](https://github.com/hoppula/refire-forum/blob/master/security-rules.bolt). I think they're now much easier to understand.

**Comment by dominictracey** [comment]
I'll pull and rebase. Thanks.

**Comment by dominictracey** [comment]
Rebased and cleared some of the Todos in security-rules.bolt.

**Comment by dominictracey** [comment]
Hmm. Now that I look back over this PR, I probably should have split off the new thread types into a different PR for you but hey ho, there it is. 😺

**Comment by hoppula** [comment]
There was a bit too much in this PR to handle at once, so I cherry picked the commits that were related to editing posts and created PR #12 with some additional changes.

I'd like to get the other new feature (starting threads with just links) in as well but I think it still needs some restructuring.

I'm planning to allow embedding Youtube videos, Imgur galleries etc. in addition to links and I'm still thinking how it'd be easiest for users to add them.

@dominictracey Could you check if #12 is ok for merge with my changes?

**Comment by dominictracey** [comment]
@hoppula that looks good and seems like a good plan. Move the YQL code out of the component layer and even consider using something like embed.ly. YQL seems to have a pretty big free quota - at least bigger than embed.ly.

**Pull request closed without merging**